jax/core.py

```diff
@@ -1150,19 +1147,20 @@ class ConcreteArray(ShapedArray):
   __slots__ = ['val']
   array_abstraction_level = 0
 
-  def __init__(self, val, weak_type=False):
-    super().__init__(np.shape(val), np.result_type(val),
-                     weak_type=weak_type)
+  def __init__(self, val, dtype, weak_type=False):
+    super().__init__(np.shape(val), dtype, weak_type=weak_type)
```

Comment:
We probably need a check that `dtype` matches `val.dtype`, or this could lead to strange things downstream

Reply:
I think it actually won't match in general though, as per [this existing comment]. Actually, that's the purpose here: we can't just read the dtype off the value (because the dtype field is the source of truth, and we don't always eagerly cast the value, e.g. while tracing, though that would be an alternative).

Reply:
I ended up removing the changes to `ConcreteArray` to make the PR a bit simpler -- I was able to just cast the value to the right type. I'm unsure of what the implications of this are, but tests seem to pass!